File: day14_part2_2017.rs

use std::fs;

fn reverse_section(arr: &mut [u8], start: usize, length: usize) {
    let n = arr.len();
    for i in 0..length / 2 {
        arr.swap((start + i) % n, (start + length - 1 - i) % n);
    }
}

fn knot_hash(input: &str) -> String {
    let mut lengths: Vec<u8> = input.bytes().collect();
    lengths.extend(&[17, 31, 73, 47, 23]);

    let mut list: Vec<u8> = (0..256).map(|x| x as u8).collect();

    let mut position = 0;
    let mut skip = 0;
    for _ in 0..64 {
        for &length in &lengths {
            reverse_section(&mut list, position, length as usize);
            position = (position + length as usize + skip) % 256;
            skip += 1;
        }
    }

    let dense_hash: Vec<u8> = list.chunks(16)
        .map(|chunk| chunk.iter().fold(0, |acc, &x| acc ^ x))
        .collect();

    let hex_hash: String = dense_hash.iter().map(|x| format!("{:02x}", x)).collect();
    hex_hash
}

fn hex_to_binary(hex_str: &str) -> String {
    let mut binary_str = String::new();
    for hex_digit in hex_str.chars() {
        let val = u8::from_str_radix(&hex_digit.to_string(), 16).unwrap();
        binary_str += &format!("{:04b}", val);
    }
    binary_str
}

fn dfs(x: usize, y: usize, grid: &mut Vec<Vec<u8>>) {
    if x >= 128 || y >= 128 || grid[x][y] != 1 {
        return;
    }
    grid[x][y] = 0;
    if x > 0 { dfs(x - 1, y, grid); }
    if x < 127 { dfs(x + 1, y, grid); }
    if y > 0 { dfs(x, y - 1, grid); }
    if y < 127 { dfs(x, y + 1, grid); }
}

fn main() {
    let data = fs::read_to_string("input.txt").expect("File reading error");
    let key_string = data.trim();

    let mut grid = vec![vec![0; 128]; 128];
    let mut total_used = 0;
    let mut regions = 0;

    for i in 0..128 {
        let row_key = format!("{}-{}", key_string, i);
        let hash = knot_hash(&row_key);
        let binary_row = hex_to_binary(&hash);

        for (j, bit) in binary_row.chars().enumerate() {
            if bit == '1' {
                grid[i][j] = 1;
                total_used += 1;
            }
        }
    }

    for i in 0..128 {
        for j in 0..128 {
            if grid[i][j] == 1 {
                regions += 1;
                dfs(i, j, &mut grid);
            }
        }
    }

    println!("{}", regions);
}

File: fsharp/day14_part2_2017.fs

open System
open System.IO
open System.Text

let reverseSection (arr:int[]) start length =
    let n = arr.Length
    for i = 0 to length/2 - 1 do
        let a = (start + i) % n
        let b = (start + length - 1 - i) % n
        let tmp = arr.[a]
        arr.[a] <- arr.[b]
        arr.[b] <- tmp

let knotHash (input:string) =
    let lengths = Array.zeroCreate (input.Length + 5)
    for i = 0 to input.Length - 1 do
        lengths.[i] <- int input.[i]
    Array.blit [|17;31;73;47;23|] 0 lengths input.Length 5
    let list = Array.init 256 id
    let mutable pos = 0
    let mutable skip = 0
    for _ in 1..64 do
        for l in lengths do
            reverseSection list pos l
            pos <- (pos + l + skip) % 256
            skip <- skip + 1
    let dense = Array.zeroCreate 16
    for i in 0..15 do
        let mutable xor = 0
        for j in 0..15 do
            xor <- xor ^^^ list.[i*16+j]
        dense.[i] <- xor
    let sb = StringBuilder(32)
    for b in dense do
        sb.AppendFormat("{0:x2}", b) |> ignore
    sb.ToString()

let hexToBinary (hexStr:string) =
    let sb = StringBuilder(128)
    for c in hexStr do
        let v = Convert.ToInt32(string c, 16)
        sb.Append(Convert.ToString(v, 2).PadLeft(4, '0')) |> ignore
    sb.ToString()

let rec dfs x y (grid:int[][]) =
    if x >= 0 && x < 128 && y >= 0 && y < 128 && grid.[x].[y] = 1 then
        grid.[x].[y] <- 0
        dfs (x-1) y grid
        dfs (x+1) y grid
        dfs x (y-1) grid
        dfs x (y+1) grid

[<EntryPoint>]
let main _ =
    let keyString = File.ReadAllText("input.txt").Trim()
    let grid = Array.init 128 (fun _ -> Array.zeroCreate 128)
    for i in 0..127 do
        let hash = knotHash (keyString + "-" + string i)
        let bits = hexToBinary hash
        for j in 0..127 do
            if bits.[j] = '1' then grid.[i].[j] <- 1
    let mutable regions = 0
    for i in 0..127 do
        for j in 0..127 do
            if grid.[i].[j] = 1 then
                regions <- regions + 1
                dfs i j grid
    printfn "%d" regions
    0
